base/abstractarray.jl

```diff
@@ -86,8 +128,48 @@ linearindexing(A::AbstractArray, B::AbstractArray...) = linearindexing(linearind
 linearindexing(::LinearFast, ::LinearFast) = LinearFast()
 linearindexing(::LinearIndexing, ::LinearIndexing) = LinearSlow()
 
+abstract IndicesBehavior
+immutable IndicesStartAt1  <: IndicesBehavior end   # indices 1:size(A,d)
+immutable IndicesUnitRange <: IndicesBehavior end   # arb UnitRange indices
+immutable IndicesList      <: IndicesBehavior end   # indices like (:cat, :dog, :mouse)
+
+indicesbehavior(A::AbstractArray) = indicesbehavior(typeof(A))
+indicesbehavior{T<:AbstractArray}(::Type{T}) = IndicesStartAt1()
+indicesbehavior(::Number) = IndicesStartAt1()
+
+abstract IndicesPerformance
+immutable IndicesFast1D <: IndicesPerformance end  # indices(A, d) is fast
+immutable IndicesSlow1D <: IndicesPerformance end  # indices(A) is better than indices(A,d)
+
+indicesperformance(A::AbstractArray) = indicesperformance(typeof(A))
+indicesperformance{T<:AbstractArray}(::Type{T}) = IndicesFast1D()
+
+"""
+    shape(A)
+
+Returns a tuple specifying the "shape" of array `A`. For arrays with
+conventional indexing (indices start at 1), this is equivalent to
+`size(A)`; otherwise it is equivalent to `incides(A)`.
+"""
```

Comment:
I've thought about this proposal a bit more. One of the things it jeopardizes is my goal of making this transition as painless as possible. Presumably we can both agree that there will be a subset of people who'd rather keep things simple and not bother worrying about supporting arrays with unconventional indexing; for those folks, one of my goals is to make it easy for them to ignore their existence and yet write code that doesn't do anything bad. "Bad" = "segfault or give a wrong answer"; I do not consider a `MethodError` (thrown because of passing an unconventional array to code that requires conventional indices) as being in the same category of badness. In such cases, a package manager can just say "a PR would be welcome" and put the burden of supporting general arrays on people who want them.

As it stands now, we can basically tell people to do a search-and-replace of `size` for `shape`, and that one step will pretty much insulate them against badness: for conventional arrays, `shape` returns the same thing as `size`, and so old code will continue to work as always. But for unconventional arrays you'll unexpectedly get a `UnitRange` and then things like `for i = 1:shape(A,1)` will throw an error because we don't have `colon(::Int, ::UnitRange)` defined. This is a safe form of failure.

If instead we eliminate `shape` and change `indices` to return a `UpTo(3)`, then making code safe requires that one switch from `size` to `indices` and then write special handling for `UpTo`. Suddenly there's much more burden on people who might just want things to keep working with minimal alteration.

An even more radical proposal would be this: get rid of both `shape` and `indices`, and roll both of their behaviors into `size`. `size(A, 1)` would act as `shape` does now: returns an `Int` or a `UnitRange` depending on the trait `indicesbehavior(A)`. We could then add the syntax `size(Int, A, 1)` to always get the size as an `Int` (returning the equivalent of `last(indices(A,1))-first(indices(A,1))+1`), and `size(Range, A, 1)` so that `size` always returns a `UnitRange`; folks writing loops could then safely say `for i in size(Range, A, 1)`. Non-updated code written as `for i = 1:size(A,1)` would safely throw a `MethodError` when passed an unconventional array.

With such a strategy, the remarkable thing is that we'd accomplish a relatively profound shakeup of the basic assumptions of julia almost invisibly. Even if `size` isn't quite the right name for this operation, this property alone makes it pretty tempting.

Of course, by having branded this release as "Arraypocalypse," we have prepared folks for relatively deep shakeups, and this transition could (if we chose) qualify as such. So it's not like I think we can't afford to break some stuff.